index.ts

```diff
@@ -269,85 +271,94 @@ export default class Analytics {
-      clearTimeout(this.timer);
-      this.timer = null;
-    }
-
-    if (!this.queue.length) {
-      this.logger.debug('queue is empty, nothing to flush');
-      setImmediate(callback);
-      return;
-    }
+  async flush(callback: AnalyticsFlushCallback = () => {}): Promise<FlushResponse | void> {
+    return await new Promise(async (resolve, reject) => {
+      // check if earlier flush was pushed to queue
+      this.logger.debug('in flush');
+
+      if (!this.enable) {
+        setImmediate(callback);
+        resolve();
+        return;
+      }
 
-    const items = this.queue.splice(0, this.flushAt);
-    const callbacks = items.map((item) => item.callback);
-    const messages = items.map((item) => {
-      // if someone mangles directly with queue
-      if (typeof item.message == 'object') {
-        item.message.sentAt = new Date();
+      if (this.timer) {
+        this.logger.debug('cancelling existing timer...');
+        clearTimeout(this.timer);
+        this.timer = null;
       }
-      return item.message;
-    });
 
-    const data = {
-      batch: messages,
-      sentAt: new Date(),
-    };
-    this.logger.debug('batch size is ' + items.length);
-    this.logger.trace('===data===', data);
+      if (!this.queue.length) {
+        this.logger.debug('queue is empty, nothing to flush');
+        setImmediate(callback);
+        resolve();
+        return;
+      }
 
-    const done = (err?: Error) => {
-      callbacks.forEach((callback_) => {
-        callback_(err);
+      const items = this.queue.splice(0, this.flushAt);
+      const callbacks = items.map((item) => item.callback);
+      const messages = items.map((item) => {
+        // if someone mangles directly with queue
+        if (typeof item.message == 'object') {
+          item.message.sentAt = new Date();
+        }
+        return item.message;
       });
-      callback(err, data);
-    };
-
-    const req = {
-      method: 'POST',
-      headers: {
-        accept: 'application/json, text/plain, */*',
-        'content-type': 'application/json;charset=utf-8',
-        'user-agent': `expo-rudder-sdk-node/${version}`,
-        authorization: 'Basic ' + Buffer.from(`${this.writeKey}:`).toString('base64'),
-      },
-      body: JSON.stringify(data),
-      timeout: this.timeout > 0 ? this.timeout : undefined,
-      retryDelay: this.getExponentialDelay.bind(this),
-      retryOn: this.isErrorRetryable.bind(this),
-    };
 
-    retryableFetch(`${this.host}`, req)
-      .then((response) => {
-        if (response.ok) {
-          done();
+      const data = {
+        batch: messages,
+        sentAt: new Date(),
+      };
+      this.logger.debug('batch size is ' + items.length);
+      this.logger.trace('===data===', data);
+
+      const done = (err?: Error) => {
+        callbacks.forEach((callback_) => {
+          callback_(err);
+        });
+        callback(err, data);
+        if (err == null) {
+          resolve({ error: undefined, data });
```

Comment:
resolving here because 
```
try {
const response = await retryableFetch(`${this.host}`, req)
// handle response
} catch {
// handle error
}
```
was only awaiting the first HTTP attempt (e.g. resolving early).

Reply:
yeah, this was a user error on my part. Not sure how I missed it, but the await syntax works just fine.